Assembling the heck out of soil - paper posted
##############################################

:author: C\. Titus Brown
:tags: ngs,python,assembly,soil
:date: 2012-12-12
:slug: 2012-assembling-the-heck-out-of-soil
:category: science


We just posted yet another pre-submission paper to arXiv.org:

**Assembling large, complex environmental metagenomes**

Authors: Adina Chuang Howe, Janet Jansson, Stephanie A. Malfatti,
Susannah Tringe, James M. Tiedje, and C. Titus Brown

`arXiv link <http://arxiv.org/abs/1212.2832>`__

`Paper repository on github <https://github.com/ged-lab/2012-assembly-methods>`__

Abstract:

   The large volumes of sequencing data required to deeply sample
   complex environments pose new challenges to sequence analysis
   approaches. De novo metagenomic assembly effectively reduces the
   total amount of data to be analyzed but requires significant
   computational resources. We apply two data reduction approaches,
   digital normalization and partitioning, to this challenge. Using a
   human gut mock community dataset, we demonstrate that these methods
   result in assemblies nearly identical to assemblies from
   unprocessed data. We then assemble two large soil metagenomes from
   matched Iowa corn and native prairie soils. The predicted
   functional content and phylogenetic origin of the assembled contigs
   indicate significant taxonomic differences despite similar
   function. The assembly strategies presented are generic and can be
   extended to any metagenome; full source code is freely available
   under a BSD license.

---

This paper combines the approaches described in our `partitioning paper (Pell et al, PNAS, 2012)
<http://pnas.org/content/early/2012/07/25/1121464109.abstract>`__,
our `digital normalization paper <http://arxiv.org/abs/1203.4802>`__,
and our `Illumina artifacts paper <http://arxiv.org/abs/1212.0159>`__
and validates them on the *de novo* assembly of a human gut "mock" community.
We then apply them to two very large soil samples.

--titus
