Some ideas for workshops and unconference models for data-intensive biology
###########################################################################

:author: C\. Titus Brown
:tags: ngs,training,workshops,unconferences
:date: 2015-03-30
:slug: 2015-some-workshop-or-unconference-ideas
:category: teaching

Here at the Lab for Data-Intensive Biology (TM) we are constantly
trying to explore new ideas for advancing the practice of biological
data sciences.  Below are some ideas that originated with or were
sharpened by conversations with Greg Wilson (Executive Director,
Software Carpentry) and Tracy Teal (Executive Director, Data
Carpentry) that I am interested in turning into reality, as part of my
`training efforts in Data Intensive Biology at UC Davis
<http://dib-training.readthedocs.org/en/pub/>`__ (also see `my blog
post <http://ivory.idyll.org/blog/2014-davis-and-training.html>`__).

Quarterly in-depth "unconferences" on advanced domain-specific data analysis
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

I spend an increasing amount of my time helping teach people `how to
analyze sequencing data
<http://ivory.idyll.org/blog/2015-a-first-workshop.html>`__, but in
practice we kinda suck at analyzing sequencing data, especially from
non-model systems.  We need some workshops to advance those effort.

I therefore am thinking of running quarterly (4x year) week-long
unconferences, each on a different topic.  The idea would be to get
together a small group of people (~15-25) actively and openly working
on various aspects of data analysis, to hang out and
collaborate/coordinate on their efforts for a week.  The goal would be
to intersperse a few presentations with lots of hacking and
communication time, with the goal of making progress on topics of
mutual interest.

The following topics are areas where I can easily imagine a week-long
technical workshop making some substantive progress:

* vet/ag genome annotation
* non-model transcriptome analysis
* geomicrobiology/function-focused metagenome analysis
* bioinformatics training (e.g. train the trainers)
* reproducibility in genomic analysis
* computational protocols and benchmarking

Importantly, these workshops would be *inexpensive* and largely
*unfunded* - I would ask participants to fund themselves, rather than
seeking to write grants to support a bunch of people.  If we can
locate them in an inexpensive place then the total cost would be in
the $1000-2000/person range, which most active research labs could
probably support.  I would seek funding for scholarships to increase
diversity of participants, but beyond that my goal would be to make
these workshops so useful that active and funded researchers *want* to
come.  (I mean, I wouldn't turn down money that dropped into my lap,
but I've had too many workshop proposals rejected as not *quite* what
the PMs wanted to get on that merry-go-round again, unless it's
critical to make something super important happen.)

One non-negotiable component for me would be that everything worked on
at these meetings would be under open licenses, and developed openly.
Although I suppose we could have a half-way house meeting where people
*interested* in opening their software could get together to do so
in a guided fashion.

These workshops would not need to be hosted at UC Davis, or by me; I
just want to make 'em happen and am happy to co-organize or
coordinate.  There are a lot of people invested in progressing on
these issues who already have some money, and so one option for
support would be to find those people and co-opt them :).

A workshop consisting of half-day focused lessons
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Last week, I ran `a workshop on starting a new project with reproducibility in mind <https://icer.msu.edu/event/reproducible-computational-analysis-%E2%80%93-how-start-new-project>`__ --

   **Description:** Computational science projects, from data analysis
   to modeling, can benefit dramatically from a little up-front
   investment in automation; starting off with version control and
   automated building of results will pay off in efficiency,
   agility, and both transparency and reproducibility of the
   results. However, most computational researchers have never been
   exposed to a completely automated analysis pipeline. I will
   demonstrate the process of initiating a new project, building a
   few initial scripts, and automating the generation of results, as
   well as building some graphs. While the topic will be from my own
   research in bioinformatics, the overall approach should apply to
   anyone doing data analysis or simulations.

   Technology used will include git, IPython Notebook, and 'make'.

   This is an interactive seminar intended for computational science
   researchers with some experience in version control and scripting
   (for example, if you've taken a Software Carpentry workshop, you
   will be at a good starting point).

This is an idea that originated with Greg - he actually nucleated the
idea in my head, I think, but then I went ahead and tried it out.
More on that workshop later, but:

We are thinking about how to do a focused series of these ~3 hour
learning opportunities, either all demo or half-demo/half
participation, for an interested group of people.  For example, my lab
could do a section on k-mer analyses of large sequencing data sets, or
on `GitHub Flow
<http://scottchacon.com/2011/08/31/github-flow.html>`__, or on
software testing, or on whatever; the important thing is that there
are tons of Software Carpentry with deep roots in one discipline or
another, and it'd be a fun way to learn from each other while teaching
to a larger audience.

This is something we might try during `the third week of our NGS
summer course
<http://ivory.idyll.org/blog/2015-summer-course-NGS.html>`__; if
you're a badged SWC instructor and want to do something on sequence
analysis, `drop me a note with a brief proposal
<mailto:ctbrown@ucdavis.edu?subject=#3rdweek%20instructor>`__.

Instructor gatherings for lesson development and testing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Tracy Teal and Jason Williams just ran `a Data Carpentry hackathon
<http://software-carpentry.org/blog/2015/01/genomics-and-assessment-hackathon.html>`__
focused on lesson development and assessment.  Riffing off of that,
what about getting instructors together to do lesson development and
testing on a regular basis, but in front of a more advanced crowd?
This would be an opportunity for people to develop and test lessons
for Software Carpentry and Data Carpentry on a tolerant audience, with
other instructors around to offer help and advice.

This is *also* something we might try during `the third week of our
NGS summer course
<http://ivory.idyll.org/blog/2015-summer-course-NGS.html>`__; if
you're a badged SWC instructor and want to do something on sequence
analysis, `please drop me a note and tell me what!
<mailto:ctbrown@ucdavis.edu?subject=#3rdweek%20instructor>`__

----

Any other thoughts on things that have worked, or might work, for advancing
training and practice in a hands-on manner?

thanks,

--titus
