An e-mail from Seven Bridges Genomics re patents and patent applications
########################################################################

:author: Deniz Kural
:tags: sbg
:date: 2016-02-29
:slug: 2016-sbg-patent-email
:category: science

Preface from Titus: this is an e-mail written by Deniz Kural of Seven
Bridges Genomics in response to concerns and accusations about their
patents and patent applications on genomics workflow engines and graph
genome analysis techniques.  It was sent to a closed list initially,
and I asked Deniz if we could publicize it; he agreed.  Note that I
removed the messages to which he was replying, given the closed nature
of the original list, and Deniz lightly edited the e-mail to remove
some personal details and fix discontinuity.

Here are some references to the Twitter conversation: `Variant
reference graphs patented!?
<https://twitter.com/drtkeane/status/697551853442179072>`__, `looks
like Seven Bridges is trying to patent Galaxy or Taverna?
<https://twitter.com/michaelhoffman/status/697555965097472000>`__

Please remember to abide by `the commenting policy
<http://ivory.idyll.org/blog/2015-site-policies.html>`__ on this blog
if you decide to post a comment here.

----

Dear All,

Firstly, thank you for the considered discussion and civility on this
thread. In this email, I've tried to cover why SBG has a patent
practice & next steps with GA4GH; and what the patents cover / prior
art concerns.

Purpose of SBGs patents:

Developed economies have a complicated relationship to patents; as one
of the few instruments underpinning the wealth these knowledge
economies create. However, science requires unfettered freedom to
build upon previous ideas, and open communication. So what to do?

Seven Bridges was founded in 2009, with our name a nod to graph theory
(Euler). We've avoided patents until 2013, and having been made aware
of various industrial and academic institutions pursuing
bioinformatics and cloud patents, have decided to pursue our own
patents as an active, operating company in a field with some publicly
traded entities aggressively prosecuting their IP claims, as a
protective measure. The patents listed in there have been submitted
before the existence of the GA4GH mailing list. Some of the government
funding we receive compels us to obtain IP and measures our success on
this criteria.  We would welcome patent reform, and would rather not
spend our funds on patents. We are significantly larger than a lab,
but tiny compared to our established software/biotech competitors, and
ultimately can't match their legal funds. We'd also love to see the
demise of patent trolls. Thus, it is not our intention to limit
scientific discussion / publication.

As an example: We've released the source code of our pipeline editor
(unrelated to graph genomes) more than a year ago under GPL, and last
month decided to also start using an Apache license to make the patent
rights of the user more explicit. We'd like to preserve an open space
for common formats and standards - common workflow language, and graph
genome exchange formats (more on this below). Open source / Apache is
a start, but not a universal panacea.

Many of the suggestions on this thread regarding patents and licenses
would be a positive step forward, and we would welcome a clarification
of GA4GH policies, including overall governance.  Although I had the
opportunity to discuss our patents & answer some questions on previous
phone calls, I also wish that we've had a wider conversation sooner.

What's patented re: Graph Genomes /  SBG patent process + prior art

We have an obligation to submit prior art, and precisely for the
reasons outlined in this thread. Likewise, it does us no good to
obtain and maintain patents that can't realistically hold
when challenged. I've tried to explain the content of our
patents and a way to incorporate community-submitted prior art below.

None of our patent applications are on "genome graphs" per se -
i.e. representing variants or sequence data as a graph genome -- we
truly believe that having a common, open representation or
specification of graph genomes is needed. This is also the basis of
our interest and participation in this group. Indeed, in my first
presentation to this group about two summers ago, I've presented an
edge graph representation, and a coordinate system that goes along
with it (which counts as a public disclosure) thus taking it off from
what can be patented.  SBG's tools must accept and write out formats
commonly used and adopted by the community to gain traction and
adoption & use open and free APIs for input and output.

We've pursued patents for improvements that go into our toolchain
(aligners, variant callers, etc.) - a good portion while it may not be
obvious at first, are tied to producing efficient / fast
implementations, having a practical benefit, and another group of
patents relate to application patents around not having restrictions
to apply graph genomes to various specific domains. Please see below
re: on prior art for a specific application or concern.

It presents an interesting challenge on how external counsel could
become a graph genome expert and write patents on this area
intelligently.  We've realized after a full cycle of this,
that external counsel often produces lower quality applications due to
lack of context and limited time investment, resulting a lot of
resources spent downstream. Thus we've been pursuing
measures to improve the quality and substance of our process,
including hiring in-house last year.

The way our patent process (essentially a chore) now works as follows:
Our attorneys sit on our weekly R&D meetings, and then follow-up if
anything novel is presented. Currently we have about 45 PhDs and 100
engineers, with 20+ R&D members working on graph-related issues alone,
and the IP activity is thus comparable to institutions of similar
size. We try not to spend engineering time on detailed patent work --
time better spent to build tools and products. Our scientists are not
trained in patent novelty vs scientific novelty, and thus tend to be
conservative with disclosures.

Thus, our researchers write a short disclosure to our attorney, who
then writes the claims (as a lawyer would), and goes back-and-forth
with external entities until a specific set of claims are finalized /
approved. For reasons beyond my understanding, the patent system seems
to encourage starting with a broad set of claims and working with the
patent examiner to make them more specific to exactly what the
engineer built.

Our attorneys are obliged to pay attention to prior art. We circulate
papers - including pioneering graph genome work, but also from highly
related fields of genome assembly and transcriptomics. It's
worth noting that many of our submissions are in "application" stage,
and indeed may be thrown out. The claims of a final submission, after
the review steps, often look very different as outlined above. We also
have the USPTO review and point out previous work. That said, more
prior art is better, and welcome.

Thus, if a person trained in reading patent claims (or anyone really),
feels like we've overlooked work from 2013 (or any of the submission
dates) & before, we'd like to hear from you on this issue.  Even
better to name which claims and applications it relates to, and which
sections of the paper.  We'll happily submit the evidence to USPTO and
remove or revise those claims. Please keep in mind sometimes we need
to wait for the USPTO to throw it out depending on which stage we're
at.  Please do not hesitate to write to ip@sbgenomics.com which goes
directly to our in house counsel.

Likewise, we'd like to have an open channel of communication in
general, so if you have any other questions or concerns please email
on related issues.

Best,
Deniz

(Titus sez: Deniz is on Twitter at `@denizkural
<https://twitter.com/denizkural>`__.)
