What I did on my summer vacation: Microbial Diversity 2016
##########################################################

:author: C\. Titus Brown
:tags: microbiology,tituslearnz
:date: 2016-08-17
:slug: 2016-summer-vacation
:category: science

I just left Woods Hole, MA, where I spent the last 6 and a half weeks
taking the `Microbial Diversity course
<http://www.mbl.edu/microbialdiversity/>`__ as a student.  It was fun,
exhausting, stimulating, and life changing!

The course had three components: a lecture series, in which
world-class microbiologists gave 2-3 hrs of talks each morning, every
Monday-Saturday of each week; a lab practical component, in which we
learned to do environmental enrichments and isolations, use
microscopes and mass specs and everything else; and a miniproject, a
self-directed project that took place the last 2-3 weeks of the
course, and for which we could make use of all of the equipment.  The
days typically went from 9am to 10pm, with a break every Sunday.

I think it's fair to say I've been microbiologized pretty thoroughly :).

The course directors were Dianne Newman and Jared Leadbetter, two
Caltech professors that I knew of old (from my own days at Caltech).
The main topic of the course was environmental microbiology and
biogeochemistry, although we did cover some biomedical topics along
with bioinformatics and genomics.  The lectures were excellent, and
there's only two or three for which I didn't take copious notes (or
live-tweet under my student account, `@tituslearnz
<https://twitter.com/tituslearnz>`__).  The other 18 students were
amazing - most were grad students, with one postdoc and two other
faculty.  The course TAs were grad students, postdocs, and staff from
the directors' labs as well as other labs.  My just-former employee
Jessica Mizzi had signed on as course coordinator, too, and it was
great to see her basically running the course on a daily basis.
Everyone was fantastic, and all in all, it was an incredibly intense
and in-depth experience.

Why Microbial Diversity?
------------------------

I applied to take the course for several reasons.  The main one is
this: even before I got involved in metagenomics at Michigan State,
I'd been moonlighting on microbial genomics for quite a while. I have
a paper on E. coli genomics, another paper on syntrophic interactions
between methane oxidizing and sulfur reducing bacteria, a third paper
on Shewanella oneidensis (with Dianne), and a fourth paper on
bone-eating worm symbionts.  While I do have a molecular biology
background from my grad work in molecular developmental biology, these
papers made me painfully aware that I knew essentially no microbial
physiology.  Yet I now work with metagenomics data regularly, and I'm
starting to get into oceanography and marine biogeochemical cycling...
I needed to know more about microbes!

On top of wanting to know more, I was also in a good position to take
6 weeks "off" to learn some new skills.  I received tenure with my
move to UC Davis, and with the Moore Foundation DDD Investigator
Award, I have some breathing room on funding.  Moreover, I have four
new postdocs coming on board (I'll be up to 6 by September!) but
they're not here yet, and most of the people in my lab were gone for
some or all of the summer.  And, finally, my wife had been invited to
Woods Hole to lecture in both Microbial Diversity and the STAMPS
course, so she would be there for three weeks.

I also expected it to be fun! I took the Embryology course back in 2005,
and it was one of the best scientific experience of my life, and it's
one of the reasons I stuck with academia.  The combination of intense
intellectualism, the general Woods Hole gemisch of awesome scientists
hanging out for the summer, and the beautiful Woods Hole weather and
beaches, make it a summer mecca for science -- so why not go back?

What were the highlights?
-------------------------

The whole thing was awesome, frankly, although by week three I was pretty
exhausted (and then it went on for another three weeks :).

One part of the experience that amazed me was how easy it was to find
new microbes.  We split into groups of 4-5 students for the first three
weeks, and each group did about 15 enrichments and isolations from the
environment, using different media and growth conditions.  (Some of these
also formed the basis for some miniprojects.)  As soon as we got critters
into pure culture, we would sequence their 16s regions and evaluate whether
or not they were already known.  From this, I personally isolated about 8
new species; some other students generated dozens of isolates.

My miniproject was super fun, too.  (You can read my final report
`here
<https://github.com/ctb/2016-micdiv-report/blob/master/2016-micdiv-report.pdf>`__
and see my powerpoint presentation `here
<https://github.com/ctb/2016-micdiv-report/blob/master/2016-micdiv-miniproject.pptx>`__.)
I set out to enrich for sulfur oxidizing autotrophs but ended up
finding some likely heterotrophs that grew anaerobically on acetate
and nitrate.  It was great fun to research media design, perform and
"debug" enrichments, and throw machines and equipment at problems (I
learned to use the anaerobic chamber, the gas chromatograph, HPLC, and
ion chromatography, along with reacquainting myself with autoclaves
and molecular biology).  I ended up finding a microbe that probably
does something relatively pedestrian (oxidizing acetate with nitrate
as an electron acceptor), with the twist that I identified the microbe
as a Shewanella species, and apparently Shewanella do not generally
metabolize acetate anaerobically (although it `has been seen before
<http://aem.asm.org/content/79/8/2818.full>`__).  I sent Jeff Gralnick
some of my isolates and he'll look into it, and maybe we'll sequence
and analyze the genomes as a follow-on.

I also got to play with Winogradsky columns! Kirsten G. and I put
together some Winogradsky columns from environmental silt from
Trunk River, a brackish marine ecosystem.  We spent a lot of time
poking them with microsensors and we're hoping to follow up on
things with some metagenomics.

More importantly for my lab's research, I got even more thoroughly
confused about the role and value of meta-omics than I had been
before.  There is an interesting gulf between how we know microbes
work in the real world (in complex and rich ecosystems) and how we
study them (frequently in isolates in the lab).  I've been edging ever
closer to working on ecosystem and molecular modeling along with data
integration of many 'omic data types, and this course really helped me
get a handle on what types of data people are generating and why.  I
also got a lot of insight into the different types of questions people
want to answer, from questions of biogeochemical cycling to resilience
and environmental adaptation to phylogeny.  I am now better prepared to
read papers across all of these areas and I'm hoping to delve deeper
into these research areas in the next decade or so.

Also, Lisa Cohen from my lab visited the course with our MinION
sequencer and we finally generated a bunch of real genomic data from
two isolates!  The genomes each resolved into two contigs, and we hope
to publish them soon. You can read `Lisa's extremely thorough blog
post on the Nanopore sequencer here
<https://monsterbashseq.wordpress.com/2016/08/13/adventures-with-ont-minion-at-mbls-microbial-diversity-course/>`__
and also see `the tutorial she put together for dealing with nanopore
data, here
<https://angus.readthedocs.io/en/2016/analyzing_nanopore_data.html>`__

A personal highlight: at the end of the course, the TAs gave me the
Tenure Deadbeat award, for "the most disturbing degeneration from faculty
to student and hopefully back." It's gonna go on my office wall.

Surely something wasn't great? Anything?
----------------------------------------

Well,

* six weeks of cafeteria food was rough.

* six weeks of dorm life was somewhat rough (ameliorated by an awesome
  roommate, Mike B. from Thea Whitman's lab).

* there was no air conditioning (really, no airflow at all) in my dorm
  room.  And it gets very hot and humid in Woods Hole at times.
  Doooooom.

  Luckily I was exhausted most of the time because otherwise getting to
  sleep would have been impossible instead of merely very difficult.

* my family was in town for three weeks, but I couldn't spend much time
  with them, and I missed them when they weren't in town!

* my e-mail inbox is now a complete and utter disaster.

...but all of these were more or less expected.

Interestingly, I ignored Twitter and much of my usual "net monitoring"
for the summer and kind of enjoyed being in my little bubble.  I doubt
I'll be able to keep it up, though.

In sum
------

* Awesome experience.
* Learned much microbiology.
* Need to learn more chemistry!


